workflow/sync/database_semaphore.go

```diff
@@ -0,0 +1,527 @@
+		And(db.Cond{controllerTimeField + " >": since})
+
+	err := session.SQL().
+		Select(stateKeyField, stateControllerField).
+		From(s.info.config.stateTable).
+		Where(db.Cond{stateNameField: s.longDbKey()}).
+		And(db.Cond{stateHeldField: false}).
+		And(db.Cond{
+			"controller IN": subquery,
+		}).
+		OrderBy(statePriorityField+" DESC", stateTimeField+" ASC").
+		All(&queue)
+
+	if err != nil {
+		s.log.WithError(err).Error("Failed to get ordered queue for semaphore notification")
+		return nil, err
+	}
+	return queue, nil
+}
+
+// notifyWaiters enqueues the next N workflows who are waiting for the semaphore to the workqueue,
+// where N is the availability of the semaphore. If semaphore is out of capacity, this does nothing.
+func (s *databaseSemaphore) notifyWaiters() {
+	limit := s.getLimit()
+	// We don't need to run a transaction here, if we get it wrong it'll right itself
+	holders, err := s.getCurrentHolders()
+	if err != nil {
+		s.log.WithError(err).Error("Failed to notify waiters")
+		return
+	}
+	holdCount := len(holders)
+
+	pending, err := s.queueOrdered(s.info.session)
+	if err != nil {
+		return
+	}
+	triggerCount := min(limit-holdCount, len(pending))
+	s.log.WithFields(log.Fields{
+		"holdCount":    holdCount,
+		"triggerCount": triggerCount,
+		"pendingCount": len(pending),
+	}).Debug("Notifying waiters for semaphore")
+	for idx := 0; idx < triggerCount; idx++ {
+		item := pending[idx]
+		if item.Controller != s.info.config.controllerName {
+			continue
+		}
+		key := workflowKey(item.Key)
+		s.log.WithFields(log.Fields{"key": item.Key, "workflowKey": key}).Debug("Enqueueing workflow for semaphore notification")
+		s.nextWorkflow(key)
+	}
+}
+
+// addToQueue adds the holderkey into priority queue that maintains the priority order to acquire the lock.
+func (s *databaseSemaphore) addToQueue(holderKey string, priority int32, creationTime time.Time) {
+	// Doesn't need a transaction, as no-one else should be inserting exactly this record ever
+	var states []stateRecord
+	err := s.info.session.SQL().
+		Select(stateKeyField).
+		From(s.info.config.stateTable).
+		Where(db.Cond{stateNameField: s.longDbKey()}).
+		And(db.Cond{stateKeyField: holderKey}).
+		And(db.Cond{stateControllerField: s.info.config.controllerName}).
+		All(&states)
+	if err != nil {
+		s.log.WithField("key", holderKey).WithError(err).Error("Failed to add to queue")
+		return
+	}
+	if len(states) > 0 {
+		return
+	}
+	record := &stateRecord{
+		Name:       s.longDbKey(),
+		Key:        holderKey,
+		Controller: s.info.config.controllerName,
+		Held:       false,
+		Priority:   priority,
+		Time:       creationTime,
+	}
+	_, err = s.info.session.Collection(s.info.config.stateTable).Insert(record)
+	switch err {
+	case nil:
+		s.log.WithField("key", holderKey).Debug("Added into queue")
+	default:
+		s.log.WithField("key", holderKey).WithError(err).Error("Failed to add to queue")
+	}
+}
+
+func (s *databaseSemaphore) removeFromQueue(holderKey string) {
```

Comment:
Same here as well, an error here does somewhat introduce a consistency issue I suppose.
On a `ReleaseAll` call node.SynchronizationStatus will be set to nil, it probably shouldn't be ?